src/types.rs

```diff
@@ -322,3 +324,21 @@ impl_error!(std::io::Error, IOError);
 impl_error!(serde_json::Error, JSON);
 impl_error!(bitcoin::hashes::hex::Error, Hex);
 impl_error!(bitcoin::consensus::encode::Error, Bitcoin);
+
+impl<T> From<std::sync::PoisonError<T>> for Error {
+    fn from(_: std::sync::PoisonError<T>) -> Self {
+        Error::CouldntLockReader
```

Comment:
I wouldn't use this error here, because `CouldntLockReader` has a very specific meaning:

> Couldn't take a lock on the reader mutex. This means that there's already another reader thread running

If a `PoisonError` happens there's not much the caller can do to recover, it has to throw away the client and recreate it. I would add an error variant that reflects this better, and write some documentation to let users know what to do in case that error happens.

Reply:
I am strongly against panics at runtime which happens not due to app design bugs, especially when you can avoid them in a small cost. What client app can do is to handle the error, terminate the tread gracefully (save the state, notify users, write debug logs, send debug info to developer) and restart the thread.

Reply:
Yeah I was suggesting to add a specific variant that reflects the true meaning of the error, and then the user can choose how to handle that variant specifically.

I generally agree with you that it's better to return an error when possible, but on the other end it's very important to ensure that the code can't accidentally end up in an inconsistent/unexpected state: in that regard, the panics on `lock()` were used to simplify things a little bit.

Basically we chose to panic rather than potentially incur in some kind of logic bug, which I still think it's a better trade-off in many cases.

Reply:
I am ok with adding error-specific variant. Or you are against removing `unwrap()`s on `lock()`?

Reply:
No that's fine, I'm just saying that now that we don't panic we have to be a bit more careful and think about the potential edge cases..